test/forcing/ForcingsEngineLumpedDataProvider_Test.cpp

```diff
@@ -0,0 +1,153 @@
+std::shared_ptr<utils::ngenPy::InterpreterUtil> TestFixture::gil_ = nullptr;
+TestFixture::provider_type* TestFixture::provider_ = nullptr;
+mpi_info TestFixture::mpi_ = {};
+
+// Initialize MPI if available, get Python GIL, and initialize forcings engine.
+void TestFixture::SetUpTestSuite()
+{
+    #if NGEN_WITH_MPI
+    MPI_Init(nullptr, nullptr);
+    #endif
+
+    TestFixture::gil_ = utils::ngenPy::InterpreterUtil::getInstance();
+
+    data_access::assert_forcings_engine_requirements();
+
+    // Create a lumped forcings engine instance
+    TestFixture::provider_ = data_access::ForcingsEngineLumpedDataProvider::lumped_instance(
+        config_file,
+        default_params.start_time,
+        default_params.end_time
+    );
+}
+
+// Destroy providers, GIL, and finalize MPI
+void TestFixture::TearDownTestSuite()
+{
+    provider_->finalize_all();
+    gil_.reset();
+
+    #if NGEN_WITH_MPI
+    PMPI_Finalize();
+    #endif
+}
+
+// ============================================================================
+
+/**
+ * Tests for the flyweight-like design of provider storage by getting
+ * a new instance of the forcings engine and verifying that it points
+ * to the same address as the static initialized `provider_` member.
+ */
+TEST_F(ForcingsEngineLumpedDataProviderTest, Storage)
+{
+    auto* inst_a = data_access::ForcingsEngineLumpedDataProvider::instance(config_file, default_params.start_time, default_params.end_time);
+    ASSERT_EQ(inst_a, provider_);
+
+    auto* inst_b = data_access::ForcingsEngineLumpedDataProvider::instance(config_file, default_params.start_time, default_params.end_time);
+    ASSERT_EQ(inst_a, inst_b);
+}
+
+TEST_F(ForcingsEngineLumpedDataProviderTest, Timing)
+{
+    EXPECT_EQ(provider_->get_data_start_time(), default_params.simulation_start_t);
+    EXPECT_EQ(provider_->get_data_stop_time(), default_params.simulation_end_t);
+    EXPECT_EQ(provider_->record_duration(), 3600); // Dependent on config file
+    EXPECT_EQ(provider_->get_ts_index_for_time(default_params.simulation_start_t), 0);
+    EXPECT_EQ(provider_->get_ts_index_for_time(default_params.simulation_end_t), 5);
+}
+
+TEST_F(ForcingsEngineLumpedDataProviderTest, VariableAccess)
+{
+    constexpr std::array<const char*, 8> expected_variables = {
+        "U2D_ELEMENT",
+        "V2D_ELEMENT",
+        "LWDOWN_ELEMENT",
+        "SWDOWN_ELEMENT",
+        "T2D_ELEMENT",
+        "Q2D_ELEMENT",
+        "PSFC_ELEMENT",
+        "RAINRATE_ELEMENT"
+    };
+
+    const auto outputs = provider_->get_available_variable_names();
+
+    ASSERT_EQ(outputs.size(), expected_variables.size());
+
+    // Check that each output variable exists in the list of expected variables
+    for (const auto& output : outputs) {
+        EXPECT_NE(
+            std::find(expected_variables.begin(), expected_variables.end(), output),
+            expected_variables.end()
```

Comment:
There's a subtle mistake here
```suggestion
    for (const auto& expected : expected_variables) {
        EXPECT_NE(
            output.find(expected),
            output.end()
```
If `outputs` consisted of N copies of a single expected variable, the test as written would have passed. We want to check that *every* expected variable is present.

Given that, (I know I'm back-tracking), do we want to assert the absence of *additional* variables? If not, then the size equality check just above could be dropped.

Reply:
Ah I noticed this too, but couldn't sort the logic to correctly check it in my head. Good catch 🙂 I think we don't need the additional assertion.